packages/yew/src/html/mod.rs

```diff
@@ -133,24 +134,151 @@ impl NodeRef {
     }
 }
 
+/// Wrapped reference to another component for later use in lifecycle methods.
+///
+/// # Example
+/// Send messages to a child component
+/// ```
+///# use yew::prelude::*;
+///
+/// struct MessageHolder {
+///     msg: String,
+/// }
+///
+/// impl Component for MessageHolder {
+///     type Message = String;
+///     type Properties = ();
+///
+///     fn create(_ctx: &Context<Self>) -> Self {
+///         Self {
+///             msg: "waiting...".to_string(),
+///         }
+///     }
+///
+///     fn update(&mut self, _ctx: &Context<Self>, message: Self::Message) -> bool {
+///         self.msg = message;
+///         true
+///     }
+///
+///     fn view(&self, _ctx: &Context<Self>) -> Html {
+///         html! { <span>{&self.msg}</span> }
+///     }
+/// }
+///
+/// pub struct Controller {
+///     log_ref: ComponentRef<MessageHolder>,
+/// }
+///
+/// impl Component for Controller {
+///     type Message = ();
+///     type Properties = ();
+///
+///     fn create(_ctx: &Context<Self>) -> Self {
+///         Self {
+///             log_ref: ComponentRef::default(),
+///         }
+///     }
+///
+///     fn view(&self, _ctx: &Context<Self>) -> Html {
+///         let onclick = {
+///             let log_ref = self.log_ref.clone();
+///             Callback::from(move |_| {
+///                 log_ref.get()
+///                     .expect("a message holder")
+///                     .send_message("example message".to_string())
+///             })
+///         };
+///         html! {
+///             <>
+///                 <MessageHolder ref={&self.log_ref} />
+///                 <button {onclick}>{"Send example message"}</button>
+///             </>
+///         }
+///     }
+/// }
+/// ```
+/// ## Relevant examples
+/// - [`nested_list`](https://github.com/yewstack/yew/tree/master/examples/nested_list)
+pub struct ComponentRef<COMP: BaseComponent>(Rc<RefCell<CompRefInner>>, PhantomData<COMP>);
```

Comment:
While editing the higher order component page, I realized that we might also want to support `forwardRef`-like feature. I'm actually playing with the idea of introducing an associated type on `BaseComponent` and defaulting it on `Component` to `Scope<Self>`, though I don't know yet how a good API to bind it would look like. For function components, maybe

```
#[function_component(ref = ImperativeHandleType)] // If specified, replace
// ImperativeHandleType becomes part of the HookContext so that type-checking use_imperative_handle works?
// Alternatively special case the name of use_imperative_handle in the macro, but that seems a bit brittle
fn Example(props: &Props) -> Html {
    use_imperative_handle(ImperativeHandleType {
        ..
    }); // doesn't this force the user to rebind it every render? debug_assert that *something* is bound?
}
```

Reply:
Introduce generics to HookContext would significantly increase code size.
(`#[hook]` cannot know the HookContext type unless they introduce a generic as well.
We don't want to take away the ability to call `use_imperative_handle` from hooks by boxing HookContext as
users may want to customise `use_imperative_handle` behaviour with custom hooks.)

How about simply do what react does?

```rust
// no need to specify type here as we can read type from the argument?
#[function_component(ref_forwarded = true)]
fn Example(props: &Props, forward_ref: ComponentRef<ImperativeHandleType>) -> Html {
    // Having a closure here so that we can delay handle creation until `effect` time.
    // ComponentRef is not usable until rendered cycle and child effects are always acknowledged before parent effects.
    // This allows concurrent mode optimisation where multiple renders can happen before it becomes rendered and we don't have to create the handle until then.
    use_imperative_handle(forward_ref, || ImperativeHandleType {
        ..
    });
}
```

Reply:
The `Component::get` (and `ComponentRef::new` to discourage use for now) methods are now only exposed for `COMP: Component`. This should allow us to cleanly transition to `use_imperative_handle` and `forwardRef` in the future without breaking. It's still generic on the component type. `NodeRef` and `ComponentRef` erase to different types (`Option<Node>` vs `Option<AnyScope>`), so I don't think merging them makes a lot of sense right now, and using the name of the component is more user-friendly than requiring to write `ComponentRef<Scope<MyComp>>`.

Reply:
> NodeRef and ComponentRef erase to different types (Option<Node> vs Option<AnyScope>)

This is an implementation detail and merging them into `Rc<dyn Any>` at anytime makes sense to me (downcasting is done by simply comparing TypeIds so it has a very low runtime cost).

> more user-friendly than requiring to write ComponentRef<Scope<MyComp>>.

This would inhibit users from writing patterns like:

```rust
let some_inner_node_ref = use_component_ref::<web_sys::Node>();

if has_touch_screen() {
    html! { <TouchTextField ref={some_inner_node_ref} /> }
} else {
    html! { <TextField ref={some_inner_node_ref} /> }
}
```

Reply:
> This would inhibit users from writing patterns like

Taking a `ref` to a component will send messages and interactions through the lifecycle and yew internals. Dropping to low-level `ref` to elements enables *directly* interacting with the DOM. I don't think it's such a good idea to blur the lines here. Wrapping an `<input>` into a component will require implementing `forwardRef` or just have a prop on the component that gets forwarded for now. With a good `forwardRef`, it should be possible to use `ref={NodeRef::default()}` unchanged.

The fact that `NodeRef` does not respect the subtyping of html is a separate matter, and maybe some more strict or generic typing there would also make sense to model the type hierarchy. But again, time for a different PR. Mixing component and element refs doesn't sit right with me, since the former might be available in SSR, the latter will not since there's no browser. High-level vs low-level, no need to mix them.

Reply:
> Taking a ref to a component will send messages and interactions through the lifecycle and yew internals. Dropping to low-level ref to elements enables directly interacting with the DOM. I don't think it's such a good idea to blur the lines here. 

Forwarding element reference to outer Component is a primary usage of `ref`.

Material UI has the following line on most components:
> The `ref` is forwarded to the root element.

In React, if you are implementing a UI library, it's (almost) expected to forward reference of elements to container Component types.

> Mixing component and element refs doesn't sit right with me, since the former might be (are?) available in SSR, 

No, refs are not available until effects are first run (`rendered()`), and effects are excluded from SSR.

Reply:
I agree with most of your points, I'm just having trouble finding the time for a good concept for unifying `NodeRef` and `ComponentRef` into one that (1) keeps the bundle size small and (2) keeps the size of the PR small. Modifying `NodeRef` touches quite a *few* places in the reconciliation, since it's ubiquitously used internally. I mean I guess I wanted to introduce an internal `DomLocation` at some point to separate `NodeRef` from the internals, but as it is right now, it's hard to unify the two concepts.

Reply:
The best way to achieve:
- (1) keeps the bundle size small and
- (2) keeps the size of the PR small

Is to not over engineering things. e.g.:

> I mean I guess I wanted to introduce an internal DomLocation at some point to separate NodeRef from the internals, but as it is right now, it's hard to unify the two concepts.

Why not simply keep NodeRef as-is (maybe `pub(crate)`) and only be converted to the new unified typed ref at the constructor of `VTag`?
This would result in a much smaller changeset.
If you are giving NodeRef a generic, you are already breaking the API so there's no incentive to rename the internal `NodeRef` and leave `NodeRef` to the public.

Having something like: `HtmlRef<T>` where T can be anything has many advantages:

1. It allows multiple handle type to be passed into the same component.
If a component's (say: `MyInput`) Handle types supports `Into<T>` for the following types, it means we can pass ref of the following to `<MyInput />`:
- `HtmlRef<HtmlInputElement>`
- `HtmlRef<HtmlElement>`
- `HtmlRef<Element>`
- `HtmlRef<EventTarget>`
- `HtmlRef<Node>`
- `HtmlRef<JsValue>`
- `HtmlRef<MyInputHandle>`
- And any other type that can be converted by `.into()`

`<input />` can accept the following type (as `HtmlInputElement` has `Into<T>` for them.):
- `HtmlRef<HtmlInputElement>`
- `HtmlRef<HtmlElement>`
- `HtmlRef<Element>`
- `HtmlRef<EventTarget>`
- `HtmlRef<Node>`
- `HtmlRef<JsValue>`

If we force a ComponentRef and a NodeRef, then it will only be able to accept:
- `ComponentRef<MyInput>`
- `NodeRef` (and can only be turned into `Node`?)
- `NodeRef<HtmlInputElement>`
   (or typed NodeRef, but then why I cannot just use `ComponentRef<HtmlInputElement>`?)
- `NodeRef<HtmlElement>`
   (conversion provided by Into<T>, but then why the same cannot be done on types of `ComponentRef<T>` if they are both provided by From and Into?)

2. It allows `HtmlRef<T>` to be shared between elements and components without introducing a new type. Which leads to smaller bundle size. Asking `T` to be `BaseComponent` inhibits this.

3. It allows the same `HtmlRef<T>` to be passed to different component / element based on which one is rendered as long as they can be converted into the target type.
(as demonstrated above.)

> a good concept for unifying NodeRef and ComponentRef into one

> The reason for unifying is that as a consumer of a forwarding component Foo, I want to write Ref<Foo>, even if that gets forwarded to some kind NodeRef, i.e. Ref<web_sys::Node>.

It's very important to state that I am not asking a unified type in this pull request.

If you want to unify NodeRef and ComponentRef, then they cannot require `T: BaseComponent` (elements are not BaseComponent).

I experimented a little bit with the concept I stated above and I think it's possible to achieve a unified `HtmlRef` type without significantly increase code weight or make big changes (even with every element correctly typed). However, this requires `ComponentRef<T>` to not be constrained to `BaseComponent`.

Without fixing all the examples and tests:

```
 5 files changed, 122 insertions(+), 41 deletions(-)
```

Reply:
Can I see the code? I literally don't understand where to `set` the "unified" storage and what the underlying stored type is. NodeRefs currently store (basically) `Option<Node>` whereas ComponentRefs store `Option<AnyScope>`. Without bending over backwards, I don't see how those can be unified.

Reply:
So, let me quickly outline how I think about refs and what I think it should look like to a user.

```
<Component ref={comp_ref} />
<input ref={el_ref} />
```

Here, `comp_ref` should have `Ref<Component>`. In such a way that if `Component` is a `forwardRef` (let me invent syntax to talk about the forwarded-to type) then this aliases to `Ref<Component::FowardedRefT>`. My problem concerns the forward being a Node, in which case this should be equivalent to `NodeRef`.

Now, you're basically proposing more strongly typed `NodeRef`s in your branch, such that `el_ref: Ref<web_sys::InputElement>`. I agree with the mission, but it's not part of this PR. All of those stay `Ref<web_sys::Node> == NodeRef` for now.

The point is this: `ref` is not part of the props. If I write `<stuff ref={stuff_ref} />`, the type of the ref (`stuff`) is always next to the `ref`. Type intuition should say that it's always possible to write `ref: Ref<stuff>`.

Reply:
> Here, comp_ref should have `Ref<Component>`
> this aliases to `Ref<Component::FowardedRefT>`
How does it work if `Component::ForwardedRefT` does not implement `BaseComponent` (element)?

> NodeRefs currently store (basically) Option<Node> whereas ComponentRefs store Option<AnyScope>.
> EDIT: found the branch on your fork :) since this only reworks NodeRef and doesn't target ComponentRef I don't think it answers my question, although yes, that's more or less how a generic NodeRef would look like.

It's not worth to implement to completion at this moment if I am not intending to file a pull request that supersedes this one. If I have also implemented `HtmlRef<T>` for component as well, I will have a complete pull request that is never going to be merged, this is counterintuitive and an inefficient way of using my time.
(Unless you agree with a complete pull request that would result in closure of this pull request.)

I am only experimenting with the part that I am not 100% sure, which is whether enabling all element types on `web_sys` would lead to a significant increase in code size or compile time. (No, it doesn't.)

This branch is simply an experiment so I can confidently say that:

(1) keeps the bundle size small and
(2) keeps the size of the PR small
(3) a good concept for unifying NodeRef and ComponentRef into one

Are all attainable at the same time and not difficult and if it's not unnecessarily over-engineered.

However, it requires `HtmlRef<T>` (`ComponentRef<T>`) to not be constrained to a certain type so any type conversion with `Into<T>` can work.

Personally, unified `Ref<T>` (where `ref` is not constrained to BaseComponent) is easier to implement (hence easy on code size), more versatile (as demonstrated above) and better matches its behaviour in React. Hence, users wouldn't be uncomfortable with the behaviour.